numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
     0,                                          /* tp_version_tag */
 };
 
+/* ______________________________________________________________________
+ * DUFunc: A call-time (hence dynamic) specializable ufunc.
+ */
+
+typedef struct {
+    PyObject_HEAD
+    PyObject      * dispatcher;
+    PyUFuncObject * ufunc;
+    PyObject      * keepalive;
+    int             frozen;
+} PyDUFuncObject;
+
+static void
+dufunc_dealloc(PyDUFuncObject *self)
+{
+    /* Note: There is no need to call PyArray_free() on
+       self->ufunc->ptr, since ufunc_dealloc() will do it for us. */
+    Py_XDECREF(self->ufunc);
+    Py_XDECREF(self->dispatcher);
+    Py_XDECREF(self->keepalive);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *
+dufunc_repr(PyDUFuncObject *dufunc)
+{
+#if PY_MAJOR_VERSION >= 3
+    return PyUnicode_FromFormat("<numba._DUFunc '%s'>", dufunc->ufunc->name);
+#else
+    return PyString_FromFormat("<numba._DUFunc '%s'>", dufunc->ufunc->name);
+#endif
+}
+
+static PyObject *
+dufunc_call(PyDUFuncObject *self, PyObject *args, PyObject *kws)
+{
+    PyObject *result=NULL;
+
+    result = PyUFunc_Type.tp_call((PyObject *)self->ufunc, args, kws);
```

Comment:
The point of dufunc_call is to act like ufunc_call unless recompilation is required.